### PR TITLE
Implement Cassandra storage method

### PR DESCRIPTION
**Pull request by XakepSDK**
This is early stage cassandra support PR. Need to test everything.

### CODE REVIEW

**Comment by XakepSDK** [comment]
3-4 days more, i'm really busy now :>

**Comment by lucko** [comment]
No worries - take your time 😄

**Comment by XakepSDK** [comment]
When you write too much sql/cql code every day:
https://puu.sh/yOciU/f21cd9b12d.png
It was little cognitive dissonance.

**Comment by XakepSDK** [comment]
@lucko can you test everything? I can't test every aspect of the plugin, because i don't know this plugin well.

File: common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/CassandraConnectionManager.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        return cluster;
+    }
+
+    @Override
+    public void close() throws Exception {
+        cluster.close();
+    }
+
+    private static boolean isNotEmpty(String str) {
+        return str != null && str.length() != 0;
```
Comment (lucko) [inline common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/CassandraConnectionManager.java]: Use Guava's `Strings#isNullOrEmpty`

File: common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/CassandraDao.java [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+                            String result = prefix.apply(sb.toString());
+                            if(!result.isEmpty()) session.execute(result);
+                            sb = new StringBuilder();
+                        }
+                    }
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
+                plugin.getLog().severe("Error occurred whilst initialising the database.");
+                shutdown();
```
Comment (lucko) [inline common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/CassandraDao.java]: Remove the `#shutdown` call.

We allow failed initialisations so if the database comes back, the plugin will start to work.

File: common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/CassandraDao.java [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+
+        TRACK_SELECT = session.prepare(this.prefix.apply("SELECT * FROM {prefix}tracks WHERE name=?"));
+        TRACK_SELECT_ALL = session.prepare(this.prefix.apply("SELECT * FROM {prefix}tracks"));
+        TRACK_INSERT = session.prepare(this.prefix.apply("INSERT INTO {prefix}tracks(name, groups) VALUES(?, ?)"));
+        TRACK_DELETE = session.prepare(this.prefix.apply("DELETE FROM {prefix}tracks WHERE name=?"));
+
+        UUID_TO_NAME_SELECT = session.prepare(this.prefix.apply("SELECT name FROM {prefix}uuid_to_name WHERE uuid=?"));
+        NAME_TO_UUID_SELECT = session.prepare(this.prefix.apply("SELECT uuid FROM {prefix}name_to_uuid WHERE name=?"));
+        UUID_TO_NAME_INSERT = session.prepare(this.prefix.apply("INSERT INTO {prefix}uuid_to_name(uuid, name) VALUES(?, ?)"));
+        NAME_TO_UUID_INSERT = session.prepare(this.prefix.apply("INSERT INTO {prefix}name_to_uuid(name, uuid) VALUES(?, ?)"));
```
Comment (lucko) [inline common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/CassandraDao.java]: I'm not really an expert on Cassandra, but certainly with the JDBC, it's considered bad practice to cache PreparedStatements in this way, as among other things, it produces race conditions when they're used concurrently.

I assume BoundStatements get around this? Idk, just wanted to check. :P

File: common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/CassandraDao.java [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+        try {
+            connectionManager.close();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Override
+    public void logAction(LogEntry entry) throws DriverException {
+        BoundStatement bind = ACTION_INSERT.bind(entry.getTimestamp(),
```
Comment (lucko) [inline common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/CassandraDao.java]: Do these need to be closed?

File: common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/CassandraDao.java [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+                entry.getActedName(),
+                entry.getAction());
+        connectionManager.getSession().execute(bind);
+    }
+
+    @Override
+    public Log getLog() throws DriverException {
+        Log.Builder builder = Log.builder();
+        BoundStatement bind = ACTION_SELECT_ALL.bind();
+        ResultSet rs = connectionManager.getSession().execute(bind);
```
Comment (lucko) [inline common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/CassandraDao.java]: Again, do ResultSets need to be closed / cleaned up?

File: common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/NodeCodec.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+    private NodeModel toInstance(UDTValue value) {
+        if(value == null) return null;
+        String permission = value.getString("id");
+        boolean enabled = value.getBool("value");
+        String server = value.getString("server");
+        String world = value.getString("world");
+        Date expiry = value.getTimestamp("expiry");
+        Map<String, String> contexts = value.getMap("contexts", String.class, String.class);
+        // Lucko, i don't like this shit.
```
Comment (lucko) [inline common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/NodeCodec.java]: Uhh? Can you elaborate a bit?

File: common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/NodeCodec.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    private UDTValue toUDTValue(NodeModel model) {
+        if(model == null) return null;
+        UDTValue udtValue = userType.newValue();
+        udtValue.setString("id", model.getPermission());
+        udtValue.setBool("value", model.getValue());
+        udtValue.setString("server", model.getServer());
+        udtValue.setString("world", model.getWorld());
+        // Lucko, i don't like this shit again.
```
Comment (lucko) [inline common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/NodeCodec.java]: ???

File: common/src/main/java/me/lucko/luckperms/common/dependencies/Dependency.java [hunk 8]
```diff
@@ -98,6 +98,30 @@
...
+            "jnr-ffi",
+            "2.0.7",
+            "lAbfqtWbw/wN6Xq/ikVecNRgwpTW9EVAfVCA25MjlkM="
+    ),
+    JNR_POSIX(
+            "com.github.jnr",
+            "jnr-posix",
+            "3.0.27",
+            "hNqwMaOGiMsuyAigEboCM8sNEUbTkCDhRHmSrVBfPXg="
+    ),
```
Comment (lucko) [inline common/src/main/java/me/lucko/luckperms/common/dependencies/Dependency.java]: Are you sure all of these are needed? Not just the Cassandra driver?

**Comment by lucko** [comment]
Thanks for your work so far.
Yes, I will try to test things out a bit before this is merged.

The nature of the project at this stage means that I'm really the sole maintainer of it. I need to have a decent understanding of how all of this works, and be convinced that it is going to be maintainable before it can be merged.

I hugely appreciate contributions of this nature, but because it will only really benefit a tiny number of users, the benefits need to outweigh the extra maintenance responsibility that comes with it.

I hope that makes sense.

**Comment by XakepSDK** [comment]
1. Driver won't start without these dependencies.
2. Ok
3. Cluster must be closed, if error occurred. Init method recall will just
recreate cluster object.
4. It's bad in cassandra to reprepare queries, it's ok
5. Nope, not required
6. That's cassandra resultset's, cleanup is unrequired
7. No method to get in unixtime
8. See 7

I need this for my project and I want to share it.

**Comment by XakepSDK** [comment]
Sooo...

**Comment by lucko** [comment]
I haven't forgotten :)

I'll get to this when I have a chance to.

**Comment by XakepSDK** [comment]
Yay, its broken now =(

**Comment by lucko** [comment]
Ok, I've just had another look through.
I'm ready to merge this when you indicate that it's ready.

Couple of final comments though...

File: common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/NodeCodec.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    private NodeModel toInstance(UDTValue value) {
+        if(value == null) return null;
+        String permission = value.getString("id");
+        boolean enabled = value.getBool("value");
+        String server = value.getString("server");
+        String world = value.getString("world");
+        Date expiry = value.getTimestamp("expiry");
+        Map<String, String> contexts = value.getMap("contexts", String.class, String.class);
```
Comment (lucko) [inline common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/NodeCodec.java]: Context sets are **not** maps.
The same key can be mapped to multiple values - they are multimaps.

For example, the following is perfectly valid:

```
{"region": ["value1", "value2"]}
```

They are better represented by `Map<String, Collection<String>>` or rather, `Map<String, Set<String>>`

File: common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/NodeCodec.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    private UDTValue toUDTValue(NodeModel model) {
+        if(model == null) return null;
+        UDTValue udtValue = userType.newValue();
+        udtValue.setString("id", model.getPermission());
+        udtValue.setBool("value", model.getValue());
+        udtValue.setString("server", model.getServer());
+        udtValue.setString("world", model.getWorld());
+        // Lucko, i don't like this shit again.
+        udtValue.setTimestamp("expiry", new Date(model.getExpiry() * 1000L));
+        return udtValue;
```
Comment (lucko) [inline common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/NodeCodec.java]: You never set the `contexts` here. See my other comment re: their handling.

The `ContextSet` cannot be safely converted to a `Map`.

File: common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/NodeCodec.java (hunk 7 above)
Comment (lucko) [inline common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/NodeCodec.java]: Please remove this comment - just totally unnecessary.

Feel free to add a helper to convert between unix-seconds and unix-millis if it really bothers you that much. :)

File: common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/NodeCodec.java (hunk 6 above)
Comment (lucko) [inline common/src/main/java/me/lucko/luckperms/common/storage/dao/cassandra/NodeCodec.java]: Remove this

**Comment by XakepSDK** [comment]
Soo, is this my error?
```
[23:07:37 ERROR]: Error occurred while enabling LuckPerms v4.0.120 (Is it up to date?)
java.lang.IncompatibleClassChangeError: Inconsistent constant pool data in classfile for class me/lucko/luckperms/lib/caffeine/cache/stats/StatsCounter. Method disabledStatsCounter()Lme/lucko/luckperms/lib/caffeine/cache/stats/StatsCounter; at index 422 is CONSTANT_MethodRef and should be CONSTANT_InterfaceMethodRef
```
Java 9

**Comment by lucko** [comment]
Yep, wasn't your fault.

Should be fixed by the latest commit(s).

**Comment by XakepSDK** [comment]
Nope, same error.
```
[01:41:37 INFO]: [LP] Loading configuration...
[01:41:37 ERROR]: Error occurred while enabling LuckPerms v4.0.124 (Is it up to date?)
java.lang.IncompatibleClassChangeError: Inconsistent constant pool data in classfile for class me/lucko/luckperms/lib/caffeine/cache/stats/StatsCounter. Method disabledStatsCounter()Lme/lucko/luckperms/lib/caffeine/cache/stats/StatsCounter; at index 422 is CONSTANT_MethodRef and should be CONSTANT_InterfaceMethodRef
	at me.lucko.luckperms.lib.caffeine.cache.Caffeine.getStatsCounterSupplier(Caffeine.java:816) ~[?:?]
	at me.lucko.luckperms.lib.caffeine.cache.UnboundedLocalCache.<init>(UnboundedLocalCache.java:69) ~[?:?]
	at me.lucko.luckperms.lib.caffeine.cache.UnboundedLocalCache$UnboundedLocalManualCache.<init>(UnboundedLocalCache.java:855) ~[?:?]
	at me.lucko.luckperms.lib.caffeine.cache.UnboundedLocalCache$UnboundedLocalLoadingCache.<init>(UnboundedLocalCache.java:917) ~[?:?]
	at me.lucko.luckperms.lib.caffeine.cache.Caffeine.build(Caffeine.java:879) ~[?:?]
	at me.lucko.luckperms.common.config.AbstractConfiguration.<init>(AbstractConfiguration.java:50) ~[?:?]
	at me.lucko.luckperms.bukkit.LPBukkitPlugin.enable(LPBukkitPlugin.java:194) ~[?:?]
	at me.lucko.luckperms.bukkit.LPBukkitPlugin.onEnable(LPBukkitPlugin.java:178) ~[?:?]
	at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:264) ~[patched_1.12.2.jar:git-Paper-1272]
	at org.bukkit.plugin.java.JavaPluginLoader.enablePlugin(JavaPluginLoader.java:316) [patched_1.12.2.jar:git-Paper-1272]
	at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:405) [patched_1.12.2.jar:git-Paper-1272]
	at org.bukkit.craftbukkit.v1_12_R1.CraftServer.enablePlugin(CraftServer.java:384) [patched_1.12.2.jar:git-Paper-1272]
	at org.bukkit.craftbukkit.v1_12_R1.CraftServer.enablePlugins(CraftServer.java:333) [patched_1.12.2.jar:git-Paper-1272]
	at net.minecraft.server.v1_12_R1.DedicatedServer.init(DedicatedServer.java:226) [patched_1.12.2.jar:git-Paper-1272]
	at net.minecraft.server.v1_12_R1.MinecraftServer.run(MinecraftServer.java:616) [patched_1.12.2.jar:git-Paper-1272]
	at java.lang.Thread.run(Thread.java:844) [?:?]
```

**Comment by lucko** [comment]
Working for me using the latest Java 9 release.

https://luck.sh/o4_1zABvDfT3WbWQM_c5wwG8cwk=

**Comment by XakepSDK** [comment]
```
user@arch ~ % java -version
openjdk version "9"
OpenJDK Runtime Environment (build 9+181)
OpenJDK 64-Bit Server VM (build 9+181, mixed mode)
```
Well, maybe i broke something? But i can't imagine, what i broke, can you check sources?

**Comment by lucko** [comment]
That's not the latest J9 version. (also I'm using the oracle release)

**Comment by XakepSDK** [comment]
Latest available in my repo. I'll try with newest java from openjdk website.

**Comment by XakepSDK** [comment]
Same error with
```
% ../jdk-9.0.4/bin/java -version          
openjdk version "9.0.4"
OpenJDK Runtime Environment (build 9.0.4+11)
OpenJDK 64-Bit Server VM (build 9.0.4+11, mixed mode)
```

**Comment by XakepSDK** [comment]
```
% ../jre-9.0.4/bin/java -version
java version "9.0.4"
Java(TM) SE Runtime Environment (build 9.0.4+11)
Java HotSpot(TM) 64-Bit Server VM (build 9.0.4+11, mixed mode)
```

My branch doesn't work, same error.

**Comment by lucko** [comment]
Try deleting the LuckPerms directory. It will need to remap the jar again.

**Comment by XakepSDK** [comment]
I need this dependencies for cassandra driver(it will work without these libs, but it will be bit slower than driver can)
https://puu.sh/z5cxI/b33fa6a1db.png

How can i add second dependency from screenshot?

**Comment by lucko** [comment]
There's a constructor in `Dependency` which will take a URL - I guess use that.

Don't worry about configuring relocations for the Cassandra depends, I'll sort that after the PR is merged.

Also, one other thing, I'd prefer if the methods added to `ImmutableContextSet` were moved to the `CassandraDao` class. API additions require a bit more thinking and consideration. It can always be added and refactored at a later time - but it's not really possible to refactor once it's in the public facing API.

**Comment by lucko** [comment]
Making a new constructor works too :)

**Comment by lucko** [comment]
I'll do some quick testing with Cassandra.

I'm having second thoughts about the NATS implementation - what are your thoughts about exposing a suitable interface in the API instead?

Obviously this would be impossible for the Storage impl (hence this PR) - but the Messaging service is much more simple.

Reason I bring this up is that it was just mentioned in Discord.

> \<redacted\> - Today at 20:56
Is there a decent way to register my own messaging system which does not use Redis with LuckPerms? Or trigger a server sync locally from the luckperms API?

> Luck - Today at 20:57
you can do the latter

> \<redacted\> - Today at 20:59
Well so far the only thing I see in the API to trigger a sync triggers the messaging system for other servers.


What d'you think?

**Comment by lucko** [comment]
I've thought previously about possibly adding RabbitMQ support, there's also things like gRPC, etc etc. I think it would be beneficial to the project to allow these to be registered via the API as opposed to being included internally.

**Comment by XakepSDK** [comment]
Api will be really cool, atm everything is hardcoded.
But this api will require things like:
1. Dependency manager api
2. Exposed config api
2. Messaging api
3. Storage api(Why not?)

**Comment by lucko** [comment]
> Storage api(Why not?)

Because the design of the storage is more prone to frequent change - whereas, the Messaging Service is just sending strings (basically).

Ok, I'm going to go ahead and implement an API for this, please can you remove the NATS part of the PR. :)

**Comment by XakepSDK** [comment]
:sad_pepe:

**Comment by XakepSDK** [comment]
Does `SetMultimap` return `Map<K, Set<V>` anyway?
```java
    @Nonnull
    public Map<String, Set<String>> toStringSetMap(ImmutableContextSet set) {
        return (Map) set.toMultimap().asMap();
    }
```
This part of code is bad, but this is one of hottest place in implementation, we can't recreate `Map` here due performance impact.

**Comment by lucko** [comment]
Yes it does. There's a method in `Multimaps` (I think) which will do a "safe" cast for you, but it's not really necessary.

**Comment by lucko** [comment]
Fyi, my initial design for the messenger API is here.

https://gist.github.com/lucko/b9aab829c8986c5a536e99379015e5d4

Subject to change, I'm going to sit and implement it in common now. Once that's done, then I'll look over the Cassandra stuff and hopefully get it merged tonight.

**Comment by lucko** [comment]
https://github.com/lucko/LuckPerms/commit/821dc4ef56cd286fcc3912f38d81f7eb0a81de5f

**Comment by lucko** [comment]
I can't get my Cassandra install to play nicely - although I suspect that's probably caused by the VM setup I have.

https://luck.sh/yhWHbhR3ZhckCFrmU7tm_lH4yo8=

Will try again tomorrow.

**Comment by XakepSDK** [comment]
Relocations broke something inside cassandra's driver ecosystem
```
[me.lucko.luckperms.lib.cassandra.Native] Could not load JNR C Library, native system calls through this library will not be available (set this logger level to DEBUG to see the full stack trace).
```
I don't like these relocations for cassandra driver. Driver has a lot bytecode+native magic inside. This can kill driver's performance.

**Comment by lucko** [comment]
Shouldn't kill performance, it just relocates all resources to a different package.

I'll try removing the relocations for the native stuff.

**Comment by XakepSDK** [comment]
I know, relocation itself does nothing bad, but driver may contain some hardcoded things for reflection, native libs may contain hardcoded object locations, its really risky to relocate everything.
Best solution - dedicated classloader for libs.

**Comment by XakepSDK** [comment]
Bump

**Comment by lucko** [comment]
Yea, you can't relocate native code.

> Best solution - dedicated classloader for libs

It's really not, lol. You then have to interact with *everything* via reflection, unless you can cast back to a standard Java interface shared by both the plugin classloader & the isolated one.

**Comment by XakepSDK** [comment]
We need to override findClass method to ignore specified classes. It's impossible with bukkit plugin system, i know this.
Can you fix relocations?

**Comment by lucko** [comment]
Yes, just sorting something else, then I'll do it.

**Comment by lucko** [comment]
Ok, I've removed the relocations for Cassandra dependencies.
To be honest, it's going to be used very rarely, so the chance of conflicts is not even worth worrying about.

I still haven't been able to get it working locally.

**Comment by XakepSDK** [comment]
Cassandra doesn't work with java 9. Try java 8.

**Comment by lucko** [comment]
Nope - still doesn't work.
Most likely because I'm testing using a VM with a bodged network configuration.

It's a bit of a problem. You said you're not able to test it completely, and I can't get an install to work locally. :/

I'm not comfortable merging this until I know it actually works.

**Comment by XakepSDK** [comment]
What cassandra version do you use? Can you connect to cassandra using
`cqlsh`?

**Comment by lucko** [comment]
```
[22:26:08] luck@luck-vm:~ $ cqlsh
Connection error: ('Unable to connect to any servers', {'127.0.0.1': error(111, "Tried connecting to [('127.0.0.1', 9042)]. Last error: Connection refused")})
```

**Comment by lucko** [comment]
```
[22:26:10] luck@luck-vm:~ $ sudo service cassandra status
● cassandra.service - LSB: distributed storage system for structured data
   Loaded: loaded (/etc/init.d/cassandra; bad; vendor preset: enabled)
   Active: active (exited) since Mon 2018-01-22 22:15:44 GMT; 10min ago
     Docs: man:systemd-sysv-generator(8)
  Process: 3092 ExecStop=/etc/init.d/cassandra stop (code=exited, status=0/SUCCESS)
  Process: 3323 ExecStart=/etc/init.d/cassandra start (code=exited, status=0/SUCCESS)
    Tasks: 0
   Memory: 0B
      CPU: 0

Jan 22 22:15:44 luck-vm systemd[1]: Starting LSB: distributed storage system for structured data...
Jan 22 22:15:44 luck-vm systemd[1]: Started LSB: distributed storage system for structured data.
```

**Comment by XakepSDK** [comment]
Cassandra isn't running.
`journalctl -u cassandra.service`
If you found something like ClassNotFoundException or something about wrong
jvm argument - you're trying to start cassandra with java 9.

**Comment by lucko** [comment]
```
[22:26:29] luck@luck-vm:~ $ journalctl -u cassandra.service
-- Logs begin at Mon 2018-01-22 22:04:15 GMT, end at Mon 2018-01-22 22:26:29 GMT. --
Jan 22 22:04:20 luck-vm systemd[1]: Starting LSB: distributed storage system for structured data...
Jan 22 22:04:20 luck-vm systemd[1]: Started LSB: distributed storage system for structured data.
Jan 22 22:08:30 luck-vm systemd[1]: Stopping LSB: distributed storage system for structured data...
Jan 22 22:08:30 luck-vm cassandra[2706]: start-stop-daemon: warning: failed to kill 2053: No such process
Jan 22 22:08:30 luck-vm systemd[1]: Stopped LSB: distributed storage system for structured data.
Jan 22 22:09:31 luck-vm systemd[1]: Starting LSB: distributed storage system for structured data...
Jan 22 22:09:31 luck-vm systemd[1]: Started LSB: distributed storage system for structured data.
Jan 22 22:10:51 luck-vm systemd[1]: Stopping LSB: distributed storage system for structured data...
Jan 22 22:10:51 luck-vm cassandra[3092]: start-stop-daemon: warning: failed to kill 2933: No such process
Jan 22 22:10:51 luck-vm systemd[1]: Stopped LSB: distributed storage system for structured data.
Jan 22 22:12:43 luck-vm systemd[1]: Stopped LSB: distributed storage system for structured data.
Jan 22 22:12:44 luck-vm systemd[1]: Stopped LSB: distributed storage system for structured data.
Jan 22 22:15:44 luck-vm systemd[1]: Starting LSB: distributed storage system for structured data...
Jan 22 22:15:44 luck-vm systemd[1]: Started LSB: distributed storage system for structured data.
```

**Comment by lucko** [comment]
Found the exception
```
ERROR [main] 2018-01-22 22:31:10,118 CassandraDaemon.java:727 - Exception encountered during startup
java.lang.AbstractMethodError: org.apache.cassandra.utils.JMXServerUtils$Exporter.exportObject(Ljava/rmi/Remote;ILjava/rmi/server/RMIClientSocketFactory;Ljava/rmi/server/RMIServerSocketFactory;Lsun/misc/Obj$
        at javax.management.remote.rmi.RMIJRMPServerImpl.export(RMIJRMPServerImpl.java:150) ~[na:1.8.0_161]
        at javax.management.remote.rmi.RMIJRMPServerImpl.export(RMIJRMPServerImpl.java:135) ~[na:1.8.0_161]
        at javax.management.remote.rmi.RMIConnectorServer.start(RMIConnectorServer.java:405) ~[na:1.8.0_161]
        at org.apache.cassandra.service.CassandraDaemon.maybeInitJmx(CassandraDaemon.java:140) [apache-cassandra-3.6.jar:3.6]
        at org.apache.cassandra.service.CassandraDaemon.setup(CassandraDaemon.java:206) [apache-cassandra-3.6.jar:3.6]
        at org.apache.cassandra.service.CassandraDaemon.activate(CassandraDaemon.java:581) [apache-cassandra-3.6.jar:3.6]
        at org.apache.cassandra.service.CassandraDaemon.main(CassandraDaemon.java:710) [apache-cassandra-3.6.jar:3.6]
```

https://mail-archives.apache.org/mod_mbox/cassandra-user/201801.mbox/%3CA6EFEA98-318F-4CD7-88FB-FED428D51B47@seeq.com%3E

https://www.mail-archive.com/user@cassandra.apache.org/msg54794.html

Downgrade it is!

**Comment by XakepSDK** [comment]
What about `/var/log/cassandra`?

**Comment by lucko** [comment]
That's where the above exception was.

**Comment by XakepSDK** [comment]
Omg, I lol'd. Microoptimisations...

**Comment by lucko** [comment]
Yeah, I really can't be bothered with this crap anymore.

I've already spent hours messing with this - I know it's not your fault but I don't have the patience to sign up for an Oracle account & downgrade my Java install *again* just to get Cassandra to work.

The amount of effort this has taken me (plus the time you've spent) so far vastly exceeds the payoff this is going to give. If we're being realistic, you're probably the only person who's actually going to use the integration.

Assuming the experience I've had with this so far is representative of the overall quality of the software, I really don't want to be in a position where I have to maintain this.

**Comment by XakepSDK** [comment]
Ah, it's so sad. It works perfectly with openjdk.(Debian 9/arch linux
tested)
Oracle isn't best choice for headless java.

**Comment by lucko** [comment]
Sorry - but the more I think about it, the more I'm tending to the opinion that this PR isn't appropriate for upstream. 

When you first opened the PR I was onboard, primarily because I assumed the change would be minimal - my understanding was that Cassandra was similar to SQL, and I (incorrectly) assumed that it would be possible to share a lot of the code with the SQL backed databases.

I'm seriously worried about the maintainability of such a change and I don't think this is something I'm going to be able to maintain. It's taken me in excess of 5 hours to just setup a demo - I'm yet to actually get the thing to run. 😅 

Also: I really just don't see it being used by anyone other than yourself... and, for that reason, I don't think the change is appropriate for the main project.


However, I really appreciate your contribution, and don't want it to be a waste of time (for either of us!)

So:

* I've established a means within the plugin to register Storage DAOs at runtime 0eba5f1cbc6b81b8f420f7987b84ea1ce3f0ab41
* This means that you can program the implementation in another plugin and provide it ready for use when LuckPerms enables

Currently, this depends on the `luckperms-common` module, and just uses the same interface as the internal DAOs. However, there's an aim to move an equivalent system into the versioned `luckperms-api` in the future.

I've setup a quick plugin here which (should) work.

https://github.com/LuckPerms/luckperms-cassandra

If you want, I'm happy to give you push access to that repository - of course you're free to do your own thing too.

Sorry for the way this has been handled. I apologise if you're annoyed - but hopefully you don't see the process as a complete waste of time. :)

**Comment by XakepSDK** [comment]
So, where can i get freshest luckperms-{anything} via maven?
Here: https://nexus.lucko.me/repository/maven-snapshots/
or here: https://nexus.lucko.me/repository/maven-releases/ ?
Thank you for time you spent for this PR.

When I have time, I will help you with these relocations.(i really don't like bytecode modification without really important reason)

**Comment by lucko** [comment]
Just reference `https://repo.lucko.me/`.

The latest builds are pulled in from Jenkins. (it mirrors from here https://ci.lucko.me/plugin/repository/everything/)

**Pull request closed without merging**